feature(proposal):  Support for Distributed Tracing (#85)
mojaloop/mojaloop-specification: fspiop-api/documents/Tracing v1.0.md

The following conventions are used in this document to identify the specified types of information.

|Type of Information|Convention|Example|
|---|---|---|
|**Elements of the API, such as resources**|Boldface|**/authorization**|
|**Variables**|Italics within curly brackets|_{ID}_|
|**Glossary terms**|Italics on first occurrence; defined in _Glossary_|The purpose of the API is to enable interoperable financial transactions between a _Payer_ (a payer of electronic funds in a payment transaction) located in one _FSP_ (an entity that provides a digital financial service to an end user) and a _Payee_ (a recipient of electronic funds in a payment transaction) located in another FSP.|
|**Library documents**|Italics|User information should, in general, not be used by API deployments; the security measures detailed in _API Signature_ and _API Encryption_ should be used instead.|

### 1.2 Document Version Information

|Version|Date|Change Description|
|---|---|---|
|**1.0**|2021-04-01|Initial version|
| | | |

## 2. Introduction

> This document details the Distributed Tracing standards to be supported for Open APIs for Interoperability (hereafter cited as the API) to ensure _tracing_ and _observability_ within the API system, and to/from API consumers.

Reviewer: As I assume that this is valid for all the APIs within Mojaloop (e.g. FSP Interoperability API, Third Party API, ..), I would suggest the below. See issue https://github.com/mojaloop/mojaloop-specification/issues/15 for more information. 

```suggestion
This document details the Distributed Tracing standards to be supported for the Mojaloop API Suite (hereafter cited as the API) to ensure _tracing_ and _observability_ within the API system, and to/from API consumers.
```

Reviewer: And another comment is that "the API" is never used after the "..hereafter cited as the API..", so the "hereafter cited as the API" can be removed.

> Distributed Tracing is a method to monitor a request end-to-end from the source to the destination, including any participating actors/components that process or forward the request. This helps to pinpoint where failures and poor performance of a specific request have occurred. As such this method is well suited to a modern micro-service architecture which includes distributed application components. In addition it also supports the inclusion of both API consumers which can be both internal and external (referred to as _vendors_). Vendors may participate either by continuing the trace, or forwarding the received tracing information without modification.

Reviewer: ```suggestion
Distributed Tracing is a method to monitor a request end-to-end from the source to the destination, including any participating actors/components that process or forward the request. This helps to pinpoint where failures and poor performance of a specific request have occurred. As such this method is well suited to a modern microservice architecture which includes distributed application components. In addition it also supports the inclusion of both API consumers which can be both internal and external (referred to as _vendors_). Vendors may participate either by continuing the trace, or forwarding the received tracing information without modification.
```

## 3. Tracing Specification

This section introduces Distributed Tracing from a Mojaloop context, including the header format, and the mechanisms used to generate, continue and forward traces.

### 3.1 Tracing Data Model

> The [Table 1](#table-1) describes the trace headers fields that are optionally included in each Interoperability API request. Tables [2](#table-2), [3](#table-3), and [4](#table-4) describes the associated values and properties that is contained by each of the respective trace header fields shown in [Table 1](#table-1).

Reviewer: To not limit this to interoperability, use just "API request" instead of:
> Interoperability API request

> **Note:** ...

Reviewer: What is the note?

Author: It has been removed.

###### Table 1

| Name | Cardinality | Type | Description | Reference |
| --- | --- | --- | --- | --- |

> | traceparent | 1 | String(55) | Incoming request containing the trace-information, including the _version_ and populated _version-format_ values.<br/>Format: _{version}_-_{version-format}_.<br/>Example: `00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01` | Sections _[3.2.2 traceparent Header Field Values](https://www.w3.org/TR/trace-context-1/#traceparent-header-field-values)_ <sup>[1](https://www.w3.org/TR/trace-context-1/#traceparent-header-field-values)</sup> |

Reviewer: "trace information" is used in the row below instead of "trace-information". I prefer "trace information".

| tracestate | 0...1 | String(512), <br/>list-member-keyvalue(0...32) | Provides optional vendor-specific trace information, and support for multiple distributed traces. It must be propagated when received, if and only if the value contains non-empty list (max of 32)  of _list-member-keyvalue_'s (Refer to [Table 4](#table-4)) that are comma (`,`) separated.<br/>Format: _{list-member-keyvalue}_,_{list-member-keyvalue}_,_{list-member-keyvalue}_. If the value of the traceparent field wasn't changed before propagation, tracestate MUST NOT be modified as well. <br/>Example: `vendorname1=opaqueValue1,vendorname2=opaqueValue2` | Section _[3.3 Tracestate Header](https://www.w3.org/TR/trace-context-1/#tracestate-header)_ <sup>[1](https://www.w3.org/TR/trace-context-1/#tracestate-header)</sup> |

> **Table 1 – Data model of HTTP header fields for Tracing**

Author: I was attempting to follow the same pattern as per the Sig doc: https://github.com/mdebarros/mojaloop-specification/blob/proposal/Distributed-Tracing-Support-for-FSPIOP-API/fspiop-api/documents/Signature%20v1.1.md

But I agree that it is cleaner. Will re-factor.

###### Table 2

| Name | Cardinality | Type | Description | Reference |
| --- | --- | --- | --- | --- |
| version | 1 | 2HEXDIGLC | Version indicator for trace-format. Refer to [Table 3](#table-3) for supported versions. | Section _[3.2.2.1 version](https://www.w3.org/TR/trace-context-1/#version)_ <sup>[1](https://www.w3.org/TR/trace-context-1/#version)</sup> |
| version-format | 1 | String(52) | A _dash_ (`-`) delimitated string included tracing information such as: trace-id, parent-id, trace-flags, etc. | Refer to [Table 3](#table-3) |
| trace-id | 1 | 32HEXDIGLC | ID representing the entire trace, i.e. the end-to-end trace identifier. | Section _[3.2.2.3 trace-id](https://www.w3.org/TR/trace-context-1/#trace-id)_ <sup>[1](https://www.w3.org/TR/trace-context-1/#trace-id)</sup> |

> | parent-id | 1 | 16HEXDIGLC | ID representing the span-id.  | Section _[3.2.2.4 parent-id](https://www.w3.org/TR/trace-context-1/#parent-id)_ <sup>[1](https://www.w3.org/TR/trace-context-1/#parent-id)</sup> |

Reviewer 2: `span-id` isn't defined anywhere. Maybe we need an image or something to help convey these different terms.

Author: Those terms are already defined in the w3 specification. I am not sure it's a good idea to replicate them here?

I.e. click the reference link and it will take you to the following:

----
3.2.2.4 parent-id
This is the ID of this request as known by the caller (in some tracing systems, this is known as `the span-id, where a span is the execution of a client request`). It is represented as an 8-byte array, for example, 00f067aa0ba902b7. All bytes as zero (0000000000000000) is considered an invalid value.

Author: I am happy to include a diagram, perhaps something similar to the following: https://docs.mojaloop.io/documentation/mojaloop-technical-overview/event-framework/#23-event-flow.

Author: I have updated table 2 to include the definition of a span for _parent-id_'s description.

| trace-flags | 1 | 2HEXDIGLC | Control flags related to the trace. | Section _[3.2.2.5 trace-flags](https://www.w3.org/TR/trace-context-1/#trace-flags)_ <sup>[1](https://www.w3.org/TR/trace-context-1/#trace-flags)</sup> |

**Table 2 – Data model for Tracing values**

###### Table 3

| Version | Format | Regex | Reference |
| --- | --- | --- | --- |
| 00 | {trace-id}_-_{parent-id}_-_{trace-flags} | `[0-9a-f]{32}-[0-9a-f]{16}-[0-9a-f]{2}` | Section _[3.2.2.2 version-format](https://www.w3.org/TR/trace-context-1/#version-format)_ <sup>[1](https://www.w3.org/TR/trace-context-1/#version-format)</sup> |

**Table 3 – Supported version-formats**

###### Table 4

| Name | Cardinality | Type | Description | Reference |
| --- | --- | --- | --- | --- |
| list-member-keyvalue | 1 | String(3...512) | Key-value pair joined by `=`, describing _vendor_ specific tracing information.<br/>Format: _{vendor-name}_=_{vendor-value}_<br/>Example: `vendorname1=opaqueValue1`  | Section _[3.3.1.3 list-members](https://www.w3.org/TR/trace-context-1/#list)_ <sup>[1](https://www.w3.org/TR/trace-context-1/#list) |

> | vendor-name | 1 | String(1-256) | Vendor (list-member) identifier which is printable lowercase ASCII which only includes `a-z`, `0-9`, underscores (`_`), dashes (`-`), asterisks (`*`), and forward slashes (`/`). | Section _[3.3.1.3.1 Key](https://www.w3.org/TR/trace-context-1/#key)_ <sup>[1](https://www.w3.org/TR/trace-context-1/#key) |

Reviewer: > String(1-256)

Should be String(1..256)

> | vendor-value | 1 | String(1-256) |  Opaque printable lowercase ASCII string excluding comma (`,`), equals (`=`), tabs, newlines, and carriage returns. | Section _[3.3.1.3.2 Value](https://www.w3.org/TR/trace-context-1/#value)_ <sup>[1](https://www.w3.org/TR/trace-context-1/#value) |

Reviewer: > String(1-256)

Should be String(1..256)

**Table 4 – Data model for _tracestate_ list member values**

> ### 3.2 Generating a Trace

Reviewer: The body text below says "To create a trace, ...", so the header should preferably say the same

```suggestion
### 3.2 Creating a Trace
```

To create the a trace, the following steps are performed:

3.2.1. Generate the _traceparent_ ([Table 1](#table-1)) header with the following values:
   > 3.2.1.1. Set the _version_ ([Table 2](#table-2)) to `00`.<br/>
   > 3.2.1.2. Generate the _trace-id_ ([Table 2](#table-2)) to represent the entire end-to-end trace.<br/>
   > 3.2.1.3. Generate a _parent-id_ ([Table 2](#table-2)) to represent a distribute portion of the trace.<br/>
   > 3.2.2.4 Set _trace-flags_ ([Table 2](#table-2)) to `01` to indicate if the trace was sampled by the source, otherwise to `00`.<br/>

3.2.2. Generate the _tracestate_ header, value below is for example purposes only as it is vendor specific: 

>    > 3.2.2.1. Add a _list-member-keyvalue_ ([Table 4](#table-4)) with value `{vendor-name}=BASE64Encoded({parent-id})`, where _vendor-name_ ([Table 4](#table-4)) is the FSP or Hub identifier and _vendor-value_ ([Table 4](#table-4)) being `BASE64Encoded({parent-id})` with the operation to base 64 encode the _parent-id_ ([Table 2](#table-2)). Base64 encoding the list member value is optional but required if its contents is not printable ASCII<sup>[1](https://www.w3.org/TR/trace-context-1/#value)</sup>.

Reviewer: Is it a requirement to use FSP identifier as vendor name? Should this not be stated in Table 4 then?

> ... where _vendor-name_ ([Table 4](#table-4)) is the FSP or Hub identifier ...

### 3.3 Continuing a Trace

> To create the a trace, the following steps are performed:

Reviewer: ```suggestion
To create a trace, the following steps are performed:
```

3.3.1. Mutate the _traceparent_ ([Table 1](#table-1)) header with the following values:
   > 3.3.1.1. _version_ ([Table 2](#table-2)) is unchanged.<br/>
   > 3.3.1.2. _trace-id_ ([Table 2](#table-2)) is unchanged.<br/>

>    > 3.3.1.3. Generate a _parent-id_ ([Table 2](#table-2)) to represent a distribute portion of the trace.<br/>

Reviewer 2: nit: should this be `to represent a distributed portion...`?

On a side not, this sentence doesn't make too much sense to me. What is a `distributed portion`?

Author: I have reworded this...let me know if this change makes more sense to you?

   > 3.3.2.4 Set _trace-flags_ ([Table 2](#table-2)) to `01` to indicate if the trace was sampled by the source, otherwise to `00`.<br/>

3.3.2. Mutate the _tracestate_ ([Table 1](#table-1)) header with the following:

>    > 3.3.2.1. Add or update (if it already exists) list member `{VENDOR}=BASE64Encoded({parent-id})`, where `BASE64Encoded({string-value})` is the operation to base 64 encode the input `string-value` to the _tracestate_ header field. Ensure that the added or updated list member is inserted/moved to the front<sup>[1](https://www.w3.org/TR/trace-context-1/#value)</sup> of the list.

Reviewer: `.. base 64 encode ..` should be "base64 encode"

   > 3.3.2.1. Ensure _tracestate_ is less than 512 characters. If necessary remove the last list members until the 512 character limit has been reached <sup>[1](https://www.w3.org/TR/trace-context-1/#tracestate-limits)</sup>.

### 3.4 Forwarding a Trace

> Vendors receiving trace request headers ([Table 1](#table-1)) do not need to participate but they MUST forward it in outgoing requests unmodified.

Reviewer: No, there should not be a requirement to forward trace headers, as that would require everyone to implement support for forwarding based on that trace headers are sent. Please see my comment https://github.com/mojaloop/mojaloop-specification/issues/71#issuecomment-741573964 on the Change Request. Trace headers must be optional, and should be decided in the scheme rules if they are used or not.

Author: Updated. I have reworded this as "recommended".

Unmodified header propagation is typically implemented in pass-through services like proxies. As such, this behavior <sup>[1](https://www.w3.org/TR/trace-context-1/#mutating-the-traceparent-field)</sup> may also be implemented in a service which currently does not collect distributed tracing information.

### 4. Examples

#### 4.1. POST /transfers request being sent by FSP1 to FSP2 through a Mojaloop Switch

> This example will result in a single end-to-end trace graph being created from `FSP1` to `FSP2` being routed through a `Mojaloop Switch` for a  **POST /transfers** linking the resulting **PUT /transfers** callback from `FSP2` to `FSP1` via the `Mojaloop Switch`.

Author: I have expanded the examples.

<br/>

> 4.1.1. `FSP1` sends a **POST /transfers** to a `Mojaloop Switch` which includes the following generated trace information:<br/><br/>

Reviewer: Can we please add some information in the example on how the trace information is generated in the `FSP1`? For example, does the tracestate t61rcWkgMzE mean something for `FSP1`?

Author: I have updated the examples to include more information on how IDs were generated.

> > traceparent: `00`-`0af7651916cd43dd8448eb211c80319c`-`b7ad6b7169203331`-`01`<br/>

Reviewer 2: nit: inconsistent use of `` here and below. Maybe we can put everything in ````, and for stuff that has changed, we can put it in `**`

e.g.:
traceparent: `00-0af7651916cd43dd8448eb211c80319c-b7ad6b7169203331-01`
then
traceparent: `00-0af7651916cd43dd8448eb211c80319c-`**00f067aa0ba902b7**`-01`

Author: Please can you be more specific about where it's not being used consistently?

I prefer the current approach.

> > tracestate: `fsp1`=`t61rcWkgMzE`<br/>

Reviewer: Please state somewhere that the FSP1 uses `fsp1` as its identifier

<br/>

> 4.1.2. The `Mojaloop Switch` receives the **POST /transfers** and responds with a `HTTP 202` accepted. The `Mojaloop Switch` continuous the trace by mutating the trace information as follows, before forwarding the the **POST /transfers** to `FSP2`:<br/><br/>

Reviewer: > continuous 

Should be continues

> > traceparent: 00-0af7651916cd43dd8448eb211c80319c-`00f067aa0ba902b7`-01<br/>

Reviewer: Please explain in the example why and how the parent-id is changed.

Author: I have expanded on table2 definition, and also included an example of how both the trace-id and parent-id are generated in the examples.

> > tracestate: `moja`=`00f067aa0ba902b7`,fsp1=t61rcWkgMzE<br/>

Reviewer: Please state somewhere that the Mojaloop Switch is using `moja` as its identifier

<br/>

> 4.1.3. `FSP2` receives the request  and responds with a `HTTP 202` accepted. The funds are committed and `FSP2` sends a **PUT /transfers** callback including the following mutated trace information to the `Mojaloop Switch`:<br/><br/>

Reviewer: ```suggestion
4.1.3. `FSP2` receives the request  and responds with a `HTTP 202` accepted. The funds are then committed and `FSP2` sends a **PUT /transfers** callback including the following mutated trace information to the `Mojaloop Switch`:<br/><br/>
```

> traceparent: 00-0af7651916cd43dd8448eb211c80319c-`b9c7c989f97918e1`-01<br/>
> tracestate: `fsp2`=`ucfJifl5GOE`,moja=00f067aa0ba902b7,fsp1=t61rcWkgMzE<br/>

<br/>
4.1.4. The `Mojaloop Switch` receives the **PUT /transfers** and responds with a `HTTP 200`. The `Mojaloop Switch` will reconstuct the trace from a combination of the _traceparent_ and _tracestate_ headers, generating a new _parent-id_ from the span-id contained in the list-member `moja`=`00f067aa0ba902b7`. The trace information will be mutated as follows, before forwarding the the **PUT /transfers** back to `FSP1`:<br/><br/>

> traceparent: 00-0af7651916cd43dd8448eb211c80319c-`53ce929d0e0e4736`-01<br/>

> > tracestate: `moja`=`53ce929d0e0e4736`,fsp2=ucfJifl5GOE,fsp1=t61rcWkgMzE<br/>

Reviewer 2: Why does `moja=00f067aa0ba902b7` get dropped here?

Author: Because a new span is created.

Author: I have re-worked the first example. Let me know if is now more clear.

<br/>

> 4.1.5. `FSP1` receives the fulfil response responding with an `HTTP 200`, and completes the transfer.

Reviewer: ```suggestion
4.1.5. `FSP1` receives the **PUT /transfers** callback and responds with an `HTTP 200`, and completes the transfer.
```

#### 4.2. POST /transfers request being sent by non-trace-participating FSP1 to non-trace-participating FSP2 through a Mojaloop Switch

This will example will result in two disparate end-to-end trace graphs being created which are NOT linked as follows:
> 4.2.a. `FSP1` to `FSP2` being routed through a `Mojaloop Switch` for a  **POST /transfers**<br/>
> 4.2.b. `FSP2` to `FSP1` being routed through a `Mojaloop Switch` for a  **PUT /transfers** callback<br/>

<br/>

> 4.2.1. `FSP1` sends a POST /transfers to a `Mojaloop Switch` which includes no trace information.

Reviewer: ```suggestion
4.2.1. `FSP1` sends a **POST /transfers** request to a `Mojaloop Switch` which includes no trace information.
```

> 4.2.2. The `Mojaloop Switch` receives the POST /transfers and responds with a HTTP 202 accepted. The `Mojaloop Switch` generates a new trace as none was found in the request headers before forwarding the the POST /transfers to `FSP2`:<br/><br/>

Reviewer: ```suggestion
4.2.2. The `Mojaloop Switch` receives the **POST /transfers** and responds with a HTTP 202 accepted. The `Mojaloop Switch` generates a new trace as none was found in the request headers before forwarding the **POST /transfers** to `FSP2`:<br/><br/>
```

> traceparent: `00`-`0af7651916cd43dd8448eb211c80319c`-`00f067aa0ba902b7`-`01`<br/>
> tracestate: `moja`=`00f067aa0ba902b7`<br/>

<br/>

> 4.2.3. `FSP2` receives the request, commits the funds and responds with a PUT /transfers callback forwarding the unmutated trace information:<br/><br/>

Reviewer: ```suggestion
4.2.3. `FSP2` receives the request and responds with a HTTP 202 accepted. `FSP2` then commits the funds and sends  a **PUT /transfers** callback forwarding the unmutated (as `FSP2` is not participating in the distributed trace) trace information:<br/><br/>
```

> traceparent: 00-0af7651916cd43dd8448eb211c80319c-b9c7c989f97918e1-01 <br/>
> tracestate: moja=00f067aa0ba902b7

<br/>

> 4.2.4. The `Mojaloop Switch` receives the PUT /transfers and responds with a HTTP 200. The `Mojaloop Switch` will reconstruct the trace from a combination of the _traceparent_ and _tracestate_ headers, generating a new _parent-id_ from the span-id contained in the list-member `moja`=`00f067aa0ba902b7`. The trace information will be mutated as follows, before forwarding the the PUT /transfers back to `FSP1`:<br/><br/>

Reviewer: Can you please explain why the Mojaloop Switch should reconstruct the trace here? Seems a bit unnecessary as it is the only participant in the trace, but I will assume that there is a good reason :)

```suggestion
4.2.4. The `Mojaloop Switch` receives the **PUT /transfers** and responds with a HTTP 200. The `Mojaloop Switch` will reconstruct the trace from a combination of the _traceparent_ and _tracestate_ headers, generating a new _parent-id_ from the span-id contained in the list-member `moja`=`00f067aa0ba902b7`. The trace information will be mutated as follows, before forwarding the **PUT /transfers** back to `FSP1`:<br/><br/>
```

Author: By this, I mean "reconstruct" the trace context within the switch. I have rewarded it as such.

It is required because each time the trace is processed by a microservices's (at a minimum - as we could even go down into internal function calls, etc), a "span" must be created to represent the work of that microservice.

Does that make it clear?

> traceparent: 00-0af7651916cd43dd8448eb211c80319c-`53ce929d0e0e4736`-01 <br/>
> tracestate: `moja`=`53ce929d0e0e4736`<br/>

<br/>

> 4.2.5. `FSP1` receives the fulfil response responding with an `HTTP 200`, and completes the transfer ignoring the trace information.

Reviewer: ```suggestion
4.2.5. `FSP1` receives the **PUT /transfers** callback and responds with an `HTTP 200`.  `FSP1` then completes the transfer. The trace information is ignored as `FSP1` is not participating in the distributed trace.
```

|Version|Date|Change Description|
|---|---|---|
|**1.0**|2021-04-01|Initial version|
| | | |

## 2. Introduction

This document details the Distributed Tracing standards to be supported for the Mojaloop API Suite to ensure _tracing_ and _observability_ within the API system, and to/from API consumers.

Distributed Tracing is a method to monitor a request end-to-end from the source to the destination, including any participating actors/components that process or forward the request. This helps to pinpoint where failures and poor performance of a specific request have occurred. As such this method is well suited to a modern microservice architecture which includes distributed application components. In addition it also supports the inclusion of both API consumers which can be both internal and external (referred to as _vendors_). Vendors may participate either by continuing the trace, or forwarding the received tracing information without modification.

## 3. Tracing Specification

This section introduces Distributed Tracing from a Mojaloop context, including the header format, and the mechanisms used to generate, continue and forward traces.

### 3.1 Tracing Data Model

> The [Table 1](#table-1-–-data-model-of-http-header-fields-for-Tracing) describes the trace headers fields that are optionally included in each API request. Tables [2](#table-2-–-data-model-for-tracing-values), [3](#table-3-–-supported-version-formats), and [4](#table-4-–-data-model-for-tracestate-list-member-values) describes the associated values and properties that is contained by each of the respective trace header fields shown in [Table 1](#table-1-–-data-model-of-http-header-fields-for-Tracing).

Reviewer: ```suggestion
[Table 1](#table-1-–-data-model-of-http-header-fields-for-Tracing) describes the trace headers fields that are optionally included in each API request. Tables [2](#table-2-–-data-model-for-tracing-values), [3](#table-3-–-supported-version-formats), and [4](#table-4-–-data-model-for-tracestate-list-member-values) describes the associated values and properties that is contained by each of the respective trace header fields shown in [Table 1](#table-1-–-data-model-of-http-header-fields-for-Tracing).
```

| Name | Cardinality | Type | Description | Reference |
| --- | --- | --- | --- | --- |
| list-member-keyvalue | 1 | String(3...512) | Key-value pair joined by `=`, describing _vendor_ specific tracing information.<br/>Format: _{vendor-name}_=_{vendor-value}_<br/>Example: `vendorname1=opaqueValue1`  | Section _[3.3.1.3 list-members](https://www.w3.org/TR/trace-context-1/#list)_ <sup>[1](https://www.w3.org/TR/trace-context-1/#list) |
| vendor-name | 1 | String(1..256) | Vendor (list-member) identifier which is printable lowercase ASCII which only includes `a-z`, `0-9`, underscores (`_`), dashes (`-`), asterisks (`*`), and forward slashes (`/`). The _FSP_ or _Hub_ identifier should be used here. | Section _[3.3.1.3.1 Key](https://www.w3.org/TR/trace-context-1/#key)_ <sup>[1](https://www.w3.org/TR/trace-context-1/#key) |
| vendor-value | 1 | String(1..256) |  Opaque printable ASCII string excluding comma (`,`), equals (`=`), tabs, newlines, and carriage returns. | Section _[3.3.1.3.2 Value](https://www.w3.org/TR/trace-context-1/#value)_ <sup>[1](https://www.w3.org/TR/trace-context-1/#value) |

### 3.2 Creating a Trace

> To create the a trace, the following steps are performed:

Reviewer: ```suggestion
To create a trace, the following steps are performed:
```

   > 3.2.1.3. Generate a _parent-id_ ([Table 2](#table-2-–-data-model-for-tracing-values)) to represent a distribute portion of the trace.<br/>
   > 3.2.2.4 Set _trace-flags_ ([Table 2](#table-2-–-data-model-for-tracing-values)) to `01` to indicate if the trace was sampled by the source, otherwise to `00`.<br/>

3.2.2. Generate the _tracestate_ header, value below is for example purposes only as it is vendor specific: 
   > 3.2.2.1. Add a _list-member-keyvalue_ ([Table 4](#table-4-–-data-model-for-tracestate-list-member-values)) with value `{vendor-name}=BASE64Encoded({parent-id})`, where _vendor-name_ ([Table 4](#table-4-–-data-model-for-tracestate-list-member-values)) is the FSP or Hub identifier and _vendor-value_ ([Table 4](#table-4-–-data-model-for-tracestate-list-member-values)) being `BASE64Encoded({parent-id})` with the operation to base64 encode the _parent-id_ ([Table 2](#table-2-–-data-model-for-tracing-values)). Base64 encoding the list member value is optional but required if its contents is not printable ASCII<sup>[1](https://www.w3.org/TR/trace-context-1/#value)</sup>.

### 3.3 Continuing a Trace

> To create a trace, the following steps are performed:

Reviewer 3: ```suggestion
To continue a trace, the following steps are performed:
```

------|-----------|----------
Payer FSP | BankNrOne | banknrone
Payee FSP | MobileMoney | mobilemoney
Mojaloop Payment Switch | Switch | moja

##### 4.1.1. Request Party Information

4.1.1.1. Request Party Information from BankNrOne to Switch

BankNrOne sends a **GET /parties** request to the Switch, and includes the following generated _traceparent_ information:

| variable | value | description |
| ---|---|--- |
| _version_ | `00` | Indicates Standard version as per [Table 3](#table-3-–-supported-version-formats) is being used. |
| _trace-id_ | `91e502e28cd723686e9940bd3f378f85` | Unique trace identifier randomly generated. Refer to [Table 2](#table-2-–-data-model-for-tracing-values). |
| _parent-id_ | `b0f903d000944947` | Updated to represent the current span. Refer to [Table 2](#table-2-–-data-model-for-tracing-values). |
| _trace-flags_ | `01 `| Indicates that BankNrOne has recorded the trace. |

The _tracestate_ is also include with the following information:
| index | _list-member_ |value | description|
| --- | --- | --- | --- |
| 0 | `banknrone` | `b0f903d000944947` | Mapped from _parent-id_. |

The following trace headers will be include in the request:

> **Note:** The fields highlighted below as **`value`** will indicate a newly generated or modified value depending on context. Non-highlighted fields will indicate an unchanged value.
> 
> > traceparent: **`00`**-**`91e502e28cd723686e9940bd3f378f85`**-**`b0f903d000944947`**-**`01`**<br/>
> > tracestate: **`banknrone`**=**`b0f903d000944947`**<br/>

Reviewer 2: nit: formatting 

```suggestion
- `traceparent`: **`00`**-**`91e502e28cd723686e9940bd3f378f85`**-**`b0f903d000944947`**-**`01`**<br/>
- `tracestate`: **`banknrone`**=**`b0f903d000944947`**<br/>

> ***Note:** The fields highlighted below as **`value`** will indicate a newly generated or modified value depending on context. Non-highlighted fields will indicate an unchanged value.*
```

Author: I am happy to make this change if the majority prefers it.

@elnyry-sam-k , @ehenrka. Your thoughts on this suggestion?

Reviewer: I'm fine with the suggestion. Two additional comments:
- The "below" in "The fields highlighted below .." should probably be changed to "in the subsequent sections" or similar, as I assume this note is for all examples?
- The **`value`** in the note is in **bold**, but values in the tables seems to be in normal (non-bold) formatting. See for example the two tables in 4.1.1.2.

<br/>
The request sample including trace headers:

```http
GET /parties/MSISDN/123456789 HTTP/1.1
Accept: application/vnd.interoperability.parties+json;version=1
Content-Type: application/vnd.interoperability.parties+json;version=1.0
Date: Tue, 15 Nov 2017 10:13:37 GMT
FSPIOP-Source: BankNrOne
traceparent: 00-91e502e28cd723686e9940bd3f378f85-b0f903d000944947-01
tracestate: banknrone=b0f903d000944947
```

4.1.1.2. Request Party Information from Switch to MobileMoney

Switch receives the request, and mutates the trace information before forwarding the request to MobileMoney.

The _traceparent_ is mutated as follows:

| variable | value | description |
| ---|---|--- |
| _version_ | 00 | Unchanged. |
| _trace-id_ | 91e502e28cd723686e9940bd3f378f85 | Unchanged as the Switch continuing the trace. |
| _parent-id_ | `b0f903d000944947` | Updated to represent the current span. Refer to [Table 2](#table-2-–-data-model-for-tracing-values). |
| _trace-flags_ | 01 | Indicates that Switch has recorded the trace. |

The _tracestate_ is mutated as follows:

| index | _list-member_ |value | description|
| --- | --- | --- | --- |

> | 1 | `moja` | `f86425c1a005e5e2` | Mapped from _parent-id_. |

Reviewer 2: Are we using 0 indexing or 1 indexing? This table is not consistent with 4.1.1.1 above

| 2 | banknrone | b0f903d000944947 | Unchanged. |

The following trace headers will be include in the request:

> **Note:** The fields highlighted below as **`value`** will indicate a newly generated or modified value depending on context. Non-highlighted fields will indicate an unchanged value.

Reviewer 2: You can remove this on subsequent examples.

tracestate: banknrone=0c3e9f44f921ca06,moja=e5687db76842fe36,mobilemoney=619d4c9d431ca708
{
    "quoteId": "7c23e80c-d078-4077-8263-2c047876fcf6",
    "transactionId": "85feac2f-39b2-491b-817e-4a03203d4f14",
    "payee": {
        "partyIdInfo": {
            "partyIdType": "MSISDN",
            "partyIdentifier": "123456789",
            "fspId": "MobileMoney"
        }
    },
    "payer": {
        "personalInfo": {
            "complexName": {
                "firstName": "Mats",
                "lastName": "Hagman"
            }
        },
        "partyIdInfo": {
            "partyIdType": "IBAN",
            "partyIdentifier": "SE4550000000058398257466",
            "fspId": "BankNrOne"
        }
    },
    "amountType": "RECEIVE",
    "amount": {
        "amount": "100",
        "currency": "USD"
    },
    "transactionType": {
        "scenario": "TRANSFER",
        "initiator": "PAYER",
        "initiatorType": "CONSUMER"
    },
    "note": "From Mats",
    "expiration": "2017-11-15T22:17:28.985-01:00"
}
```

4.1.2.2. Request Quote from Switch to MobileMoney

Switch receives the request, and mutates the trace information before forwarding the request to MobileMoney.

> **Note:** See previous section [4.2.1. Request Party Information](#421-request-party-information) for more detail.

Reviewer: The link seems to be incorrect, I would guess that it should be to 4.1.1.2. Request Party Information from Switch to MobileMoney?

Author: Fixed. It's meant to be 4.1.1.

{
    "quoteId": "7c23e80c-d078-4077-8263-2c047876fcf6",
    "transactionId": "85feac2f-39b2-491b-817e-4a03203d4f14",
    "payee": {
        "partyIdInfo": {
            "partyIdType": "MSISDN",
            "partyIdentifier": "123456789",
            "fspId": "MobileMoney"
        }
    },
    "payer": {
        "personalInfo": {
            "complexName": {
                "firstName": "Mats",
                "lastName": "Hagman"
            }
        },
        "partyIdInfo": {
            "partyIdType": "IBAN",
            "partyIdentifier": "SE4550000000058398257466",
            "fspId": "BankNrOne"
        }
    },
    "amountType": "RECEIVE",
    "amount": {
        "amount": "100",
        "currency": "USD"
    },
    "transactionType": {
        "scenario": "TRANSFER",
        "initiator": "PAYER",
        "initiatorType": "CONSUMER"
    },
    "note": "From Mats",
    "expiration": "2017-11-15T22:17:28.985-01:00"
}
```

4.1.2.3. Quote response callback from MobileMoney to Switch

MobileMoney receives the request, and mutates the trace information before responding with the **PUT /quotes** callback to the Switch.

> **Note:** See previous section [4.2.1. Request Party Information](#421-request-party-information) for more detail.

Reviewer: The link seems to be incorrect, I would guess that it should be to 4.1.1.2. Request Party Information from Switch to MobileMoney?

Author: Fixed. It's meant to refer to "[4.1.1. Request Party Information](#411-request-party-information)" in general.

thcmxzc29uIg0KICAgICAgICAgICAgfQ0KICAgICAgICB9DQogICAgfSwNCiAgICAicGF5ZXIi-
OiB7DQogICAgICAgICJwZXJzb25hbEluZm8iOiB7DQogICAgICAgICAgICAiY29tcGxleE5hbWUi-
OiB7DQogICAgICAgICAgICAgICAgImZpcnN0TmFtZSI6ICJNYXRzIiwNCiAgICAgICAgICAgICAgI-
CAibGFzdE5hbWUiOiAiSGFnbWFuIg0KICAgICAgICAgICAgfQ0KICAgICAgICB9LA0KICAgICAgI-
CAicGFydHlJZEluZm8iOiB7DQogICAgICAgICAgICAicGFydHlJZFR5cGUiOiAiSUJBTiIsDQogICAgI-
CAgICAgICAicGFydHlJZGVudGlmaWVyI-
jogIlNFNDU1MDAwMDAwMDA1ODM5ODI1NzQ2NiIsDQogICAgICAgICAgICAiZnNwSWQiOiAiQmFua05yT25
lIg0KICAgICAgICB9DQogICAgfSwNCiAgICAiYW1vdW50Ijogew0KICAgICAgICAiYW1vdW50IjogIjEw-
MCIsDQogICAgICAgICJjdXJyZW5jeSI6ICJVU0QiDQogICAgfSwNCiAgICAidHJhbnNhY3Rpb25UeXBlI-
jogew0KICAgICAgICAic2NlbmFyaW8iOiAiVFJBTlNGRVIiLA0KICAgICAgICAiaW5pdGlhdG9yI-
jogIlBBWUVSIiwNCiAgICAgICAgImluaXRpYXRvclR5cGUiOiAiQ09OU1VNRVIiDQogICAgfSwNCiAgI-
CAibm90ZSI6ICJGcm9tIE1hdHMiDQp9DQo\u003d\u003d",
    "condition": "fH9pAYDQbmoZLPbvv3CSW2RfjU4jvM4ApG_fqGnR7Xs"
}
```

4.1.2.4. Quote response callback from Switch to BankNrOne

Switch receives the request, and mutates the trace information before forwarding the **PUT /quotes** callback to BankNrOne.

> **Note:** See previous section [4.2.1. Request Party Information](#421-request-party-information) for more detail.

Reviewer: The link seems to be incorrect, I would guess that it should be to 4.1.1.2. Request Party Information from Switch to MobileMoney?

Author: Fixed. It's meant to refer to "4.1.1. Request Party Information" in general.

OiB7DQogICAgICAgICAgICAgICAgImZpcnN0TmFtZSI6ICJNYXRzIiwNCiAgICAgICAgICAgICAgI-
CAibGFzdE5hbWUiOiAiSGFnbWFuIg0KICAgICAgICAgICAgfQ0KICAgICAgICB9LA0KICAgICAgI-
CAicGFydHlJZEluZm8iOiB7DQogICAgICAgICAgICAicGFydHlJZFR5cGUiOiAiSUJBTiIsDQogICAgI-
CAgICAgICAicGFydHlJZGVudGlmaWVyI-
jogIlNFNDU1MDAwMDAwMDA1ODM5ODI1NzQ2NiIsDQogICAgICAgICAgICAiZnNwSWQiOiAiQmFua05yT25
lIg0KICAgICAgICB9DQogICAgfSwNCiAgICAiYW1vdW50Ijogew0KICAgICAgICAiYW1vdW50IjogIjEw-
MCIsDQogICAgICAgICJjdXJyZW5jeSI6ICJVU0QiDQogICAgfSwNCiAgICAidHJhbnNhY3Rpb25UeXBlI-
jogew0KICAgICAgICAic2NlbmFyaW8iOiAiVFJBTlNGRVIiLA0KICAgICAgICAiaW5pdGlhdG9yI-
jogIlBBWUVSIiwNCiAgICAgICAgImluaXRpYXRvclR5cGUiOiAiQ09OU1VNRVIiDQogICAgfSwNCiAgI-
CAibm90ZSI6ICJGcm9tIE1hdHMiDQp9DQo\u003d\u003d",
    "condition": "fH9pAYDQbmoZLPbvv3CSW2RfjU4jvM4ApG_fqGnR7Xs"
}
```

4.1.2.5. Quote processed by BankNrOne

The BankNrOne receives the callback request with the necessary information to request a transfer.

##### 4.1.3 Request Transfer

4.1.3.1. Request Transfer from BankNrOne to Switch

BankNrOne sends a **POST /transfers** request to the Switch while continuing the trace `91e502e28cd723686e9940bd3f378f85`.

> **Note:** See previous section [4.2.1. Request Party Information](#421-request-party-information) for more detail.

Reviewer: The link seems to be incorrect, I would guess that it should be to 4.1.1.2. Request Party Information from Switch to MobileMoney?

Author: Fixed. It's meant to refer to "4.1.1. Request Party Information" in general.

CAgICAgICAiZmlyc3ROYW1lIjogIkhlbnJpayIsDQogICAgICAgICAgICAgICAgImxhc3ROYW1lIjogIk-
thcmxzc29uIg0KICAgICAgICAgICAgfQ0KICAgICAgICB9DQogICAgfSwNCiAgICAicGF5ZXIi-
OiB7DQogICAgICAgICJwZXJzb25hbEluZm8iOiB7DQogICAgICAgICAgICAiY29tcGxleE5hbWUi-
OiB7DQogICAgICAgICAgICAgICAgImZpcnN0TmFtZSI6ICJNYXRzIiwNCiAgICAgICAgICAgICAgI-
CAibGFzdE5hbWUiOiAiSGFnbWFuIg0KICAgICAgICAgICAgfQ0KICAgICAgICB9LA0KICAgICAgI-
CAicGFydHlJZEluZm8iOiB7DQogICAgICAgICAgICAicGFydHlJZFR5cGUiOiAiSUJBTiIsDQogICAgI- CAgICAgICAicGFydHlJZGVudGlmaWVyI-
jogIlNFNDU1MDAwMDAwMDA1ODM5ODI1NzQ2NiIsDQogICAgICAgICAgICAiZnNwSWQiOiAiQmFua05yT25 lIg0KICAgICAgICB9DQogICAgfSwNCiAgICAiYW1vdW50Ijogew0KICAgICAgICAiYW1vdW50IjogIjEw-
MCIsDQogICAgICAgICJjdXJyZW5jeSI6ICJVU0QiDQogICAgfSwNCiAgICAidHJhbnNhY3Rpb25UeXBlI-
jogew0KICAgICAgICAic2NlbmFyaW8iOiAiVFJBTlNGRVIiLA0KICAgICAgICAiaW5pdGlhdG9yI-
jogIlBBWUVSIiwNCiAgICAgICAgImluaXRpYXRvclR5cGUiOiAiQ09OU1VNRVIiDQogICAgfSwNCiAgI-
CAibm90ZSI6ICJGcm9tIE1hdHMiDQp9DQo\u003d\u003d",
"condition": "fH9pAYDQbmoZLPbvv3CSW2RfjU4jvM4ApG_fqGnR7Xs" 
}
```

4.1.3.2. Request Transfer from Switch to MobileMoney

Switch receives the request, and mutates the trace information before forwarding the request to MobileMoney.

> **Note:** See previous section [4.2.1. Request Party Information](#421-request-party-information) for more detail.

Reviewer: The link seems to be incorrect, I would guess that it should be to 4.1.1.2. Request Party Information from Switch to MobileMoney?

Author: Fixed. It's meant to refer to "4.1.1. Request Party Information" in general.

thcmxzc29uIg0KICAgICAgICAgICAgfQ0KICAgICAgICB9DQogICAgfSwNCiAgICAicGF5ZXIi-
OiB7DQogICAgICAgICJwZXJzb25hbEluZm8iOiB7DQogICAgICAgICAgICAiY29tcGxleE5hbWUi-
OiB7DQogICAgICAgICAgICAgICAgImZpcnN0TmFtZSI6ICJNYXRzIiwNCiAgICAgICAgICAgICAgI-
CAibGFzdE5hbWUiOiAiSGFnbWFuIg0KICAgICAgICAgICAgfQ0KICAgICAgICB9LA0KICAgICAgI-
CAicGFydHlJZEluZm8iOiB7DQogICAgICAgICAgICAicGFydHlJZFR5cGUiOiAiSUJBTiIsDQogICAgI- CAgICAgICAicGFydHlJZGVudGlmaWVyI-
jogIlNFNDU1MDAwMDAwMDA1ODM5ODI1NzQ2NiIsDQogICAgICAgICAgICAiZnNwSWQiOiAiQmFua05yT25 lIg0KICAgICAgICB9DQogICAgfSwNCiAgICAiYW1vdW50Ijogew0KICAgICAgICAiYW1vdW50IjogIjEw-
MCIsDQogICAgICAgICJjdXJyZW5jeSI6ICJVU0QiDQogICAgfSwNCiAgICAidHJhbnNhY3Rpb25UeXBlI-
jogew0KICAgICAgICAic2NlbmFyaW8iOiAiVFJBTlNGRVIiLA0KICAgICAgICAiaW5pdGlhdG9yI-
jogIlBBWUVSIiwNCiAgICAgICAgImluaXRpYXRvclR5cGUiOiAiQ09OU1VNRVIiDQogICAgfSwNCiAgI-
CAibm90ZSI6ICJGcm9tIE1hdHMiDQp9DQo\u003d\u003d",
"condition": "fH9pAYDQbmoZLPbvv3CSW2RfjU4jvM4ApG_fqGnR7Xs" 
}
```

4.1.3.3. Transfer response callback from MobileMoney to Switch

MobileMoney receives the request, and mutates the trace information before responding with the **PUT /transfers** callback to the Switch.

> **Note:** See previous section [4.2.1. Request Party Information](#421-request-party-information) for more detail.

Reviewer: The link seems to be incorrect, I would guess that it should be to 4.1.1.2. Request Party Information from Switch to MobileMoney?

Author: Fixed. It's meant to refer to "4.1.1. Request Party Information" in general.

**Note:** The fields highlighted below as **`value`** will indicate a newly generated or modified value depending on context. Non-highlighted fields will indicate an unchanged value.

> traceparent: 00-91e502e28cd723686e9940bd3f378f85-**`da6dbe3af96ed639`**-01<br/>
> tracestate: **`mobilemoney`**=**`da6dbe3af96ed639`**,moja=bb68d98a0840aa99,banknrone=82f942e2566117c0<br/>

<br/>
The request sample including trace headers:

```http
PUT /transfers/11436b17-c690-4a30-8505-42a2c4eafb9d HTTP/1.1
Content-Type: application/vnd.interoperability.transfers+json;version=1.0
Content-Length: 166
Date: Tue, 15 Nov 2017 10:14:02 GMT
FSPIOP-Source: MobileMoney
FSPIOP-Destination: BankNrOne
traceparent: 00-91e502e28cd723686e9940bd3f378f85-da6dbe3af96ed639-01
tracestate: mobilemoney=da6dbe3af96ed639,moja=bb68d98a0840aa99,banknrone=82f942e2566117c0
{
    "fulfilment": "mhPUT9ZAwd-BXLfeSd7-YPh46rBWRNBiTCSWjpku90s",
    "completedTimestamp": "2017- 11-16T04:15:35.513+01:00",
    "transferState": "COMMITTED"
}
```

4.1.3.4. Transfer response callback from Switch to BankNrOne

Switch receives the request, and mutates the trace information before forwarding the **PUT /transfers** callback to BankNrOne.

> **Note:** See previous section [4.2.1. Request Party Information](#421-request-party-information) for more detail.

Reviewer: The link seems to be incorrect, I would guess that it should be to 4.1.1.2. Request Party Information from Switch to MobileMoney?

Author: Fixed. It's meant to refer to "4.1.1. Request Party Information" in general.
